src/promptflow/promptflow/_utils/version_hint_utils.py

```diff
@@ -0,0 +1,117 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import datetime
+import json
+import sys
+import contextlib
+
+from promptflow._constants import (LAST_HINT_TIME, LAST_CHECK_TIME, PF_VERSION_CHECK, CLI_PACKAGE_NAME,
+                                   HINT_FREQUENCY_DAY, GET_PYPI_FREQUENCY_DAY, LATEST_VERSION, CURRENT_VERSION)
+from promptflow._sdk._constants import HOME_PROMPT_FLOW_DIR
+
+
+if sys.platform.startswith("win"):
+    import msvcrt
+else:
+    import fcntl
+
+HINT_ACTIVITY_NAME = ["pf.flows.test", "pf.runs.create_or_update", "pfazure.flows.create_or_update",
+                      "pfazure.runs.create_or_update"]
+
+
+@contextlib.contextmanager
+def acquire_lock(filename):
+    if not sys.platform.startswith("win"):
+        with open(filename, "a+") as f:
+            fcntl.flock(f, fcntl.LOCK_EX)
+            yield f
+            fcntl.flock(f, fcntl.LOCK_UN)
+            f.close()
+    else:  # Windows
+        with open(filename, "w") as f:
+            msvcrt.locking(f.fileno(), msvcrt.LK_LOCK, 1)
+            yield f
+            msvcrt.locking(f.fileno(), msvcrt.LK_UNLCK, 1)
+            f.close()
+
+
+def get_cached_versions():
+    from promptflow._sdk._utils import read_write_by_user
+    lock_path = HOME_PROMPT_FLOW_DIR / (PF_VERSION_CHECK + '.lock')
+    with acquire_lock(lock_path):
+        (HOME_PROMPT_FLOW_DIR / PF_VERSION_CHECK).touch(mode=read_write_by_user(), exist_ok=True)
+        with open(HOME_PROMPT_FLOW_DIR / PF_VERSION_CHECK, "r") as f:
+            try:
+                cached_versions = json.load(f)
+            except json.decoder.JSONDecodeError:
+                cached_versions = {}
+        return cached_versions
+
+
+def dump_cached_versions(cached_versions):
+    lock_path = HOME_PROMPT_FLOW_DIR / (PF_VERSION_CHECK + '.lock')
+    with acquire_lock(lock_path):
+        with open(HOME_PROMPT_FLOW_DIR / PF_VERSION_CHECK, "w") as f:
+            json.dump(cached_versions, f)
+
+
+def get_latest_version_from_pypi(package_name):
+    pypi_url = f"https://pypi.org/pypi/{package_name}/json"
+    try:
+        import requests
+        response = requests.get(pypi_url)
+        if response.status_code == 200:
```

Comment:
what's our behavior if we gets stucked here?

Reply:
will added timeout, then it will just return when get pypi url failed